### DISCUSSION TITLE
Is there a way to model agents in 3D Space using mesa-geo?

### DISCUSSION THREAD

**Discussion by ifoxfoot** [mesa/mesa-geo#137] [Q&A]
Hello,

I just started using mesa and mesa-geo and I'm really enjoying it so far! 

I'm currently modeling reef systems, with the eventual goal of modeling reef size, height, roughness, etc. I've approached this where each reef builder is an agent, which is born, grows, dies, then decays. So I have modeled dimensions at the agent level. The problem is that at the reef level, reef agents are stacked up on each other. Eventually the ones below start to die and decay. If I am interested in modeling how big and bumpy a reef is....I need 3 dimensions.

How could I go about solving this problem? I'm thinking each generation could be a layer, and the top layer reef agent would have to "look down" to find how many reef builders are in the below layers, get their dimensions, and then calculate its height? To make things even more complicated, the surface where the reefs are built aren't flat to begin with 🤦‍♀️ ...I'm thinking I need a raster for this? 

Any suggestions would be most welcome. Thanks!!

**Comment by wang-boyu** [comment]
Hi! Thank you for using Mesa & Mesa-Geo!

For 2-D raster model you might want to check the [Rainfall model](https://github.com/projectmesa/mesa-geo/blob/main/examples/rainfall). To compare with yours:
  - Like your reef agent, rain drops at each location accumulate and have heights.
  - Unlike your reef agent, rain drops move around based on the accumulated heights (i.e., a [digital elevation model](https://en.wikipedia.org/wiki/Digital_elevation_model)).

Since it's a 2-D environment, it is essentially a top-down view on the ground. For your model, however, this would only visualize the top, living layer of reef agents. They will all look the same, unless you color code each raster cell based on its height, like what was done in the Rainfall model.

I think a 3-D environment will be in-general very useful. There is a ticket in Mesa that talked about this years ago: https://github.com/projectmesa/mesa/pull/402 and @jackiekazil was looking for examples (https://github.com/projectmesa/mesa/pull/402#issuecomment-355459035). Any thoughts? @rht @jackiekazil @tpike3

**Reply by ifoxfoot** [reply]
Hi! Thanks for the fast reply!

And thanks for directing me to the rainfall example. I had seen it before but looking closer, I think it kinda gets at what I want to do and will be a really helpful start.

I don't really care about the visualization as much as I care about the output. I think a raster of the reef height would be a good output. The problem is that reef agents will continue to grow and decay even when they are not on the top layer, so I need to somehow keep track of the below layers...eek! Reef agents will also grow horizontally so as they grow they might end up with multiple reef agents of different sizes on top of them. So they will have to expand into to the nearest cells.

I will continue to think about this. If I come up with a solution I will respond to this thread so that others can find it. If you all implement some 3D capability let me know! I'm sure it would be very useful for a lot of folks 🙏

**Reply by rht** [reply]
#402 is a 3D rendering of a 2D grid, like you would see in Hash.ai rendering. There is an N dimensional grid discussion at https://github.com/projectmesa/mesa/discussions/1500 (cc @Tortar). We have a lot of huge grid refactors for Mesa 2.0 that I think the N-D grid will only happen after the refactor. Though, upvoting [#1500](https://github.com/projectmesa/mesa/discussions/1500) should give us a signal how badly needed the feature is.

**Reply by wang-boyu** [reply]
@ifoxfoot this looks like an interesting agent-based model. Good luck!

@rht https://github.com/projectmesa/mesa/discussions/1500 is a more generic solution. With this I suppose we still need some separate visualization modules (both in Python and JS) to render 3D spaces? For example, modules like `Gird3DVisualization.py` and corresponding `Grid3DModule.js`, that uses 3D js libraries such as three.js. This was what was similarly done in https://github.com/projectmesa/mesa/pull/402, which could be a good starting point.

**Reply by wang-boyu** [reply]
I just realized that this is a mesa-geo discussion - 3D rendering on maps will be different from that in mesa even if it's implemented. Will need to somehow integrate leaflet.js or use some other js libraries in mesa-geo.

**Reply by rht** [reply]
I second @ifoxfoot that having a 3D/N-D grid object implemented may come first before having to design their visualization module.
